taichi/runtime/llvm/runtime.cpp

```diff
@@ -1105,12 +1108,21 @@ void cpu_parallel_range_for(Context *context,
 void gpu_parallel_range_for(Context *context,
                             int begin,
                             int end,
-                            RangeForTaskFunc *func) {
+                            range_for_xlogue prologue,
+                            RangeForTaskFunc *func,
+                            range_for_xlogue epilogue,
+                            const std::size_t tls_size) {
   int idx = thread_idx() + block_dim() * block_idx() + begin;
+  alignas(8) char tls_buffer[tls_size];
```

Comment:
See if I understand this correctly. 

* `tls_buffer` is still a thread local buffer, pretty much like adstack? (Previously I thought TLS buffer was `__shared__`.)
* There is no warp level reduction in this PR yet. The saving of global atomics purely comes from grid-strided loops? I.e. if each thread still only handles one coordinate, then there's no difference?
* In the epilogue, we will do the warp-level reduction on these TLS buffers. By that time we need `__synchthreads()` to make sure all works in `body` are done.

Reply:
> * `tls_buffer` is still a thread local buffer, pretty much like adstack? (Previously I thought TLS buffer was `__shared__`.)

“Thread local” storage and shared memory (“block local”) storage are actually two different types of storage. On CUDA and Metal we need to support both.
 1. TLS is allocated as thread registers
 2. Block local storage (shared memory) is allocated per block

I should change the name to `Local Storage`.




> * There is no warp level reduction in this PR yet. The saving of global atomics purely comes from grid-strided loops? I.e. if each thread still only handles one coordinate, then there's no difference?

Right, I find that we can achieve 95% peak without warp-level reductions in this test case. You are right, we should try to add wrap-level reduction in the future - that would be helpful for smaller arrays.



> * In the epilogue, we will do the warp-level reduction on these TLS buffers. By that time we need `__synchthreads()` to make sure all works in `body` are done.

Actually, if that's just TLS + warp-level reduction, then `__synthreads()` is not needed. If you do use shared memory (which is not in this PR), then we need `__synthreads()`.